django/db/models/sql/query.py

```diff
@@ -508,20 +454,32 @@ def get_aggregation(self, using, added_aggregate_names):
                         annotation_mask |= inner_query.annotations[name].get_refs()
                     inner_query.set_annotation_mask(annotation_mask)
 
-            relabels = {t: "subquery" for t in inner_query.alias_map}
-            relabels[None] = "subquery"
             # Remove any aggregates marked for reduction from the subquery
-            # and move them to the outer AggregateQuery.
-            col_cnt = 0
+            # and move them to the outer AggregateQuery. This requires making
+            # sure all columns referenced by the aggregates are selected in the
+            # subquery. It is achieved by retrieving all column references from
+            # the aggregates, explicitly selecting them if they are not already,
+            # and making sure the aggregates are repointed to referenced to them.
+            col_refs = {}
             for alias, expression in list(inner_query.annotation_select.items()):
+                if not expression.is_summary:
+                    continue
                 annotation_select_mask = inner_query.annotation_select_mask
-                if expression.is_summary:
-                    expression, col_cnt = inner_query.rewrite_cols(expression, col_cnt)
-                    outer_query.annotations[alias] = expression.relabeled_clone(
-                        relabels
-                    )
-                    del inner_query.annotations[alias]
-                    annotation_select_mask.remove(alias)
+                replacements = {}
+                for col in self._gen_cols([expression], resolve_refs=False):
+                    if not (col_ref := col_refs.get(col)):
+                        index = len(col_refs) + 1
```

Comment:
The +1 here was to have the exact same SQL as `rewrite_cols` generated (starting `__col%d` at 1 instead of 0) but is purely cosmetic.